### PR TITLE
fixes #21043; fixes a named exception in the infixAs expression which generate an implicit uninitialized let statement

### LINKED ISSUE #21043: Named except clauses and experimental strictDefs don't work together.
### Description

This exception syntax doesn't compile with the new definite assignment analysis for let.

```nim
{.experimental: "strictDefs".}

proc foo =
  try:
    discard
  except ValueError as e:
    echo(e.msg)

foo()
```

### Nim Version

Nim Compiler Version 1.7.3 [Linux: amd64]
Compiled at 2022-12-07
Copyright (c) 2006-2022 by Andreas Rumpf

git hash: d6b485493d712a20155a3e77f44c19c34e47ea73
active boot switches: -d:release --gc:markAndSweep

### Current Output

```text
temp.nim(5, 10) Error: 'let' symbol requires an initialization
```

### Expected Output

```text
Compiles
```

### Possible Solution

_No response_

### Additional Information

Regression after https://github.com/nim-lang/Nim/pull/21024

### PR DESCRIPTION
**Pull request by ringabout**
fixes #21043

### CODE REVIEW

**Comment by ringabout** [comment]
Before

```nim
func main =
  try:
    discard 123
  except ValueError as e:
    discard e.msg

main()
```
It compiles before but shouldn't because it implicitly calls the `getCurrentException` function, which has side effects. To keep backwards compatibility, I use `cast(noSideEffect)` to make it continue to compile.

File: compiler/sempass2.nim [hunk 1]
```diff
@@ -332,7 +332,7 @@ proc useVar(a: PEffects, n: PNode) =
   if a.inExceptOrFinallyStmt > 0:
     incl s.flags, sfUsedInFinallyOrExcept
   if isLocalSym(a, s):
-    if sfNoInit in s.flags:
+    if {sfNoInit, sfImportc} * s.flags != {}:
```
Comment (ringabout) [inline compiler/sempass2.nim:335]:
```nim
{.experimental: "strictDefs".}

proc main =
  var
    SOMAXCONN {.importc, header: "winsock2.h".}: cint
  echo SOMAXCONN
```
Without this patch, the code above continues to fail.

**Comment by Araq** [comment]
Doing this transformation in the sem phase is fundamentally wrong because then macros see the expanded code and not `except E as e` anymore!

**Comment by ringabout** [comment]
Where should I do the transform ref https://github.com/nim-lang/Nim/pull/21046#issuecomment-1342827522? Otherwise, should I lift the `let` to a global scope (not sure how to do that)?

**Comment by ringabout** [comment]
Sure, I will fix it => https://github.com/nim-lang/Nim/pull/21051/commits/bc23eec7bdc73d5cdc4dbe968053047c3368687a I thought I cannot use `addDecl(c, symbol)` before the let assignment.

**Comment by ringabout** [comment]
@Araq Btw, what's your opinion of this code which implicitly generates side effects `getCurrentException`?

```nim
func main =
  try:
    discard 123
  except ValueError as e:
    discard e.msg

main()
```

**Comment by Araq** [comment]
> Btw, what's your opinion of this code which implicitly generates side effects getCurrentException?

It doesn't generate a side effect. The fact that exceptions are implemented via thread local storage is an implementation detail that MUST NOT interfere with Nim's side effect analysis.

**Comment by ringabout** [comment]
Succeeded by https://github.com/nim-lang/Nim/pull/21081

**Pull request closed without merging**